File: test/src/dense_qp_solve.cpp
//
// Copyright (c) 2022 INRIA
//
#include <iostream>
#include <doctest.hpp>
#include <Eigen/Core>
#include <Eigen/Cholesky>
#include <proxsuite/proxqp/dense/dense.hpp>
#include <proxsuite/linalg/veg/util/dbg.hpp>
#include <proxsuite/proxqp/utils/random_qp_problems.hpp>

using T = double;
using namespace proxsuite;
using namespace proxsuite::proxqp;

DOCTEST_TEST_CASE("proxqp::dense: test init with fixed sizes matrices")
{
  double sparsity_factor = 0.15;
  T eps_abs = T(1e-9);
  utils::rand::set_seed(1);
  dense::isize dim = 10;

  dense::isize n_eq(5), n_in(2);
  T strong_convexity_factor(1.e-2);
  proxqp::dense::Model<T> qp = proxqp::utils::dense_strongly_convex_qp(
    dim, n_eq, n_in, sparsity_factor, strong_convexity_factor);

  Eigen::Matrix<T, 10, 10> H = qp.H;
  Eigen::Matrix<T, 10, 1> g = qp.g;
  Eigen::Matrix<T, 5, 10> A = qp.A;
  Eigen::Matrix<T, 5, 1> b = qp.b;
  Eigen::Matrix<T, 2, 10> C = qp.C;
  Eigen::Matrix<T, 2, 1> l = qp.l;
  Eigen::Matrix<T, 2, 1> u = qp.u;

  {
    Results<T> results = dense::solve<T>(
      H, g, A, b, C, l, u, nullopt, nullopt, nullopt, eps_abs, 0);

    T pri_res = std::max((qp.A * results.x - qp.b).lpNorm<Eigen::Infinity>(),
                         (helpers::positive_part(qp.C * results.x - qp.u) +
                          helpers::negative_part(qp.C * results.x - qp.l))
                           .lpNorm<Eigen::Infinity>());
    T dua_res = (qp.H * results.x + qp.g + qp.A.transpose() * results.y +
                 qp.C.transpose() * results.z)
                  .lpNorm<Eigen::Infinity>();
    DOCTEST_CHECK(pri_res <= eps_abs);
    DOCTEST_CHECK(dua_res <= eps_abs);

    std::cout << "------using API solving qp with dim: " << dim
              << " neq: " << n_eq << " nin: " << n_in << std::endl;
    std::cout << "primal residual: " << pri_res << std::endl;
    std::cout << "dual residual: " << dua_res << std::endl;
    std::cout << "total number of iteration: " << results.info.iter
              << std::endl;
    std::cout << "setup timing " << results.info.setup_time << " solve time "
              << results.info.solve_time << std::endl;
  }

  {
    dense::QP<T> qp_problem(dim, n_eq, 0);
    qp_problem.init(H, g, A, b, nullopt, nullopt, nullopt);
    qp_problem.settings.eps_abs = eps_abs;
    qp_problem.solve();

    const Results<T>& results = qp_problem.results;

    T pri_res = (qp.A * results.x - qp.b).lpNorm<Eigen::Infinity>();
    T dua_res = (qp.H * results.x + qp.g + qp.A.transpose() * results.y)
                  .lpNorm<Eigen::Infinity>();
    DOCTEST_CHECK(pri_res <= eps_abs);
    DOCTEST_CHECK(dua_res <= eps_abs);

    std::cout << "------using API solving qp with dim: " << dim
              << " neq: " << n_eq << " nin: " << n_in << std::endl;
    std::cout << "primal residual: " << pri_res << std::endl;
    std::cout << "dual residual: " << dua_res << std::endl;
    std::cout << "total number of iteration: " << results.info.iter
              << std::endl;
    std::cout << "setup timing " << results.info.setup_time << " solve time "
              << results.info.solve_time << std::endl;
  }
}

DOCTEST_TEST_CASE("sparse random strongly convex qp with equality and "
                  "inequality constraints: test solve function")
{

  std::cout << "---testing sparse random strongly convex qp with equality and "
               "inequality constraints: test solve function---"
            << std::endl;
  double sparsity_factor = 0.15;
  T eps_abs = T(1e-9);
  utils::rand::set_seed(1);
  dense::isize dim = 10;

  dense::isize n_eq(dim / 4);
  dense::isize n_in(dim / 4);
  T strong_convexity_factor(1.e-2);
  proxqp::dense::Model<T> qp = proxqp::utils::dense_strongly_convex_qp(
    dim, n_eq, n_in, sparsity_factor, strong_convexity_factor);
  Results<T> results = dense::solve<T>(qp.H,
                                       qp.g,
                                       qp.A,
                                       qp.b,
                                       qp.C,
                                       qp.l,
                                       qp.u,
                                       nullopt,
                                       nullopt,
                                       nullopt,
                                       eps_abs,
                                       0);

  T pri_res = std::max((qp.A * results.x - qp.b).lpNorm<Eigen::Infinity>(),
                       (helpers::positive_part(qp.C * results.x - qp.u) +
                        helpers::negative_part(qp.C * results.x - qp.l))
                         .lpNorm<Eigen::Infinity>());
  T dua_res = (qp.H * results.x + qp.g + qp.A.transpose() * results.y +
               qp.C.transpose() * results.z)
                .lpNorm<Eigen::Infinity>();
  DOCTEST_CHECK(pri_res <= eps_abs);
  DOCTEST_CHECK(dua_res <= eps_abs);

  std::cout << "------using API solving qp with dim: " << dim
            << " neq: " << n_eq << " nin: " << n_in << std::endl;
  std::cout << "primal residual: " << pri_res << std::endl;
  std::cout << "dual residual: " << dua_res << std::endl;
  std::cout << "total number of iteration: " << results.info.iter << std::endl;
  std::cout << "setup timing " << results.info.setup_time << " solve time "
            << results.info.solve_time << std::endl;
}

DOCTEST_TEST_CASE("sparse random strongly convex qp with equality and "
                  "inequality constraints: test solve with different rho value")
{

  std::cout << "---testing sparse random strongly convex qp with equality and "
               "inequality constraints: test solve with different rho value---"
            << std::endl;
  double sparsity_factor = 0.15;
  T eps_abs = T(1e-9);
  utils::rand::set_seed(1);
  dense::isize dim = 10;

  dense::isize n_eq(dim / 4);
  dense::isize n_in(dim / 4);
  T strong_convexity_factor(1.e-2);
  proxqp::dense::Model<T> qp = proxqp::utils::dense_strongly_convex_qp(
    dim, n_eq, n_in, sparsity_factor, strong_convexity_factor);
  Results<T> results = dense::solve<T>(qp.H,
                                       qp.g,
                                       qp.A,
                                       qp.b,
                                       qp.C,
                                       qp.l,
                                       qp.u,
                                       nullopt,
                                       nullopt,
                                       nullopt,
                                       eps_abs,
                                       0,
                                       T(1.E-7));
  DOCTEST_CHECK(results.info.rho == T(1.E-7));
  T pri_res = std::max((qp.A * results.x - qp.b).lpNorm<Eigen::Infinity>(),
                       (helpers::positive_part(qp.C * results.x - qp.u) +
                        helpers::negative_part(qp.C * results.x - qp.l))
                         .lpNorm<Eigen::Infinity>());
  T dua_res = (qp.H * results.x + qp.g + qp.A.transpose() * results.y +
               qp.C.transpose() * results.z)
                .lpNorm<Eigen::Infinity>();
  DOCTEST_CHECK(pri_res <= eps_abs);
  DOCTEST_CHECK(dua_res <= eps_abs);

  std::cout << "------using API solving qp with dim: " << dim
            << " neq: " << n_eq << " nin: " << n_in << std::endl;
  std::cout << "primal residual: " << pri_res << std::endl;
  std::cout << "dual residual: " << dua_res << std::endl;
  std::cout << "total number of iteration: " << results.info.iter << std::endl;
  std::cout << "setup timing " << results.info.setup_time << " solve time "
            << results.info.solve_time << std::endl;
}

DOCTEST_TEST_CASE(
  "sparse random strongly convex qp with equality and "
  "inequality constraints: test solve with different mu_eq and mu_in values")
{

  std::cout << "---testing sparse random strongly convex qp with equality and "
               "inequality constraints: test solve with different mu_eq and "
               "mu_in values---"
            << std::endl;
  double sparsity_factor = 0.15;
  T eps_abs = T(1e-9);
  utils::rand::set_seed(1);
  dense::isize dim = 10;

  dense::isize n_eq(dim / 4);
  dense::isize n_in(dim / 4);
  T strong_convexity_factor(1.e-2);
  proxqp::dense::Model<T> qp = proxqp::utils::dense_strongly_convex_qp(
    dim, n_eq, n_in, sparsity_factor, strong_convexity_factor);
  Results<T> results = dense::solve<T>(qp.H,
                                       qp.g,
                                       qp.A,
                                       qp.b,
                                       qp.C,
                                       qp.l,
                                       qp.u,
                                       nullopt,
                                       nullopt,
                                       nullopt,
                                       eps_abs,
                                       0,
                                       nullopt,
                                       T(1.E-2),
                                       T(1.E-2));
  T pri_res = std::max((qp.A * results.x - qp.b).lpNorm<Eigen::Infinity>(),
                       (helpers::positive_part(qp.C * results.x - qp.u) +
                        helpers::negative_part(qp.C * results.x - qp.l))
                         .lpNorm<Eigen::Infinity>());
  T dua_res = (qp.H * results.x + qp.g + qp.A.transpose() * results.y +
               qp.C.transpose() * results.z)
                .lpNorm<Eigen::Infinity>();
  DOCTEST_CHECK(pri_res <= eps_abs);
  DOCTEST_CHECK(dua_res <= eps_abs);

  std::cout << "------using API solving qp with dim: " << dim
            << " neq: " << n_eq << " nin: " << n_in << std::endl;
  std::cout << "primal residual: " << pri_res << std::endl;
  std::cout << "dual residual: " << dua_res << std::endl;
  std::cout << "total number of iteration: " << results.info.iter << std::endl;
  std::cout << "setup timing " << results.info.setup_time << " solve time "
            << results.info.solve_time << std::endl;
}

DOCTEST_TEST_CASE("sparse random strongly convex qp with equality and "
                  "inequality constraints: test warm starting")
{

  std::cout << "---testing sparse random strongly convex qp with equality and "
               "inequality constraints: test warm starting---"
            << std::endl;
  double sparsity_factor = 0.15;
  T eps_abs = T(1e-9);
  utils::rand::set_seed(1);
  dense::isize dim = 10;

  dense::isize n_eq(dim / 4);
  dense::isize n_in(dim / 4);
  T strong_convexity_factor(1.e-2);
  proxqp::dense::Model<T> qp = proxqp::utils::dense_strongly_convex_qp(
    dim, n_eq, n_in, sparsity_factor, strong_convexity_factor);
  auto x_wm = utils::rand::vector_rand<T>(dim);
  auto y_wm = utils::rand::vector_rand<T>(n_eq);
  auto z_wm = utils::rand::vector_rand<T>(n_in);
  Results<T> results = dense::solve<T>(
    qp.H, qp.g, qp.A, qp.b, qp.C, qp.l, qp.u, x_wm, y_wm, z_wm, eps_abs, 0);
  T pri_res = std::max((qp.A * results.x - qp.b).lpNorm<Eigen::Infinity>(),
                       (helpers::positive_part(qp.C * results.x - qp.u) +
                        helpers::negative_part(qp.C * results.x - qp.l))
                         .lpNorm<Eigen::Infinity>());
  T dua_res = (qp.H * results.x + qp.g + qp.A.transpose() * results.y +
               qp.C.transpose() * results.z)
                .lpNorm<Eigen::Infinity>();
  DOCTEST_CHECK(pri_res <= eps_abs);
  DOCTEST_CHECK(dua_res <= eps_abs);

  std::cout << "------using API solving qp with dim: " << dim
            << " neq: " << n_eq << " nin: " << n_in << std::endl;
  std::cout << "primal residual: " << pri_res << std::endl;
  std::cout << "dual residual: " << dua_res << std::endl;
  std::cout << "total number of iteration: " << results.info.iter << std::endl;
  std::cout << "setup timing " << results.info.setup_time << " solve time "
            << results.info.solve_time << std::endl;
}

DOCTEST_TEST_CASE("sparse random strongly convex qp with equality and "
                  "inequality constraints: test verbose = true")
{

  std::cout << "---testing sparse random strongly convex qp with equality and "
               "inequality constraints: test verbose = true ---"
            << std::endl;
  double sparsity_factor = 0.15;
  T eps_abs = T(1e-9);
  utils::rand::set_seed(1);
  dense::isize dim = 10;

  dense::isize n_eq(dim / 4);
  dense::isize n_in(dim / 4);
  T strong_convexity_factor(1.e-2);
  proxqp::dense::Model<T> qp = proxqp::utils::dense_strongly_convex_qp(
    dim, n_eq, n_in, sparsity_factor, strong_convexity_factor);
  bool verbose = true;
  Results<T> results = dense::solve<T>(qp.H,
                                       qp.g,
                                       qp.A,
                                       qp.b,
                                       qp.C,
                                       qp.l,
                                       qp.u,
                                       nullopt,
                                       nullopt,
                                       nullopt,
                                       eps_abs,
                                       0,
                                       nullopt,
                                       nullopt,
                                       nullopt,
                                       verbose);
  T pri_res = std::max((qp.A * results.x - qp.b).lpNorm<Eigen::Infinity>(),
                       (helpers::positive_part(qp.C * results.x - qp.u) +
                        helpers::negative_part(qp.C * results.x - qp.l))
                         .lpNorm<Eigen::Infinity>());
  T dua_res = (qp.H * results.x + qp.g + qp.A.transpose() * results.y +
               qp.C.transpose() * results.z)
                .lpNorm<Eigen::Infinity>();
  DOCTEST_CHECK(pri_res <= eps_abs);
  DOCTEST_CHECK(dua_res <= eps_abs);

  std::cout << "------using API solving qp with dim: " << dim
            << " neq: " << n_eq << " nin: " << n_in << std::endl;
  std::cout << "primal residual: " << pri_res << std::endl;
  std::cout << "dual residual: " << dua_res << std::endl;
  std::cout << "total number of iteration: " << results.info.iter << std::endl;
  std::cout << "setup timing " << results.info.setup_time << " solve time "
            << results.info.solve_time << std::endl;
}

DOCTEST_TEST_CASE("sparse random strongly convex qp with equality and "
                  "inequality constraints: test no initial guess")
{

  std::cout << "---testing sparse random strongly convex qp with equality and "
               "inequality constraints: test no initial guess ---"
            << std::endl;
  double sparsity_factor = 0.15;
  T eps_abs = T(1e-9);
  utils::rand::set_seed(1);
  dense::isize dim = 10;

  dense::isize n_eq(dim / 4);
  dense::isize n_in(dim / 4);
  T strong_convexity_factor(1.e-2);
  proxqp::dense::Model<T> qp = proxqp::utils::dense_strongly_convex_qp(
    dim, n_eq, n_in, sparsity_factor, strong_convexity_factor);
  InitialGuessStatus initial_guess = InitialGuessStatus::NO_INITIAL_GUESS;
  Results<T> results = dense::solve<T>(qp.H,
                                       qp.g,
                                       qp.A,
                                       qp.b,
                                       qp.C,
                                       qp.l,
                                       qp.u,
                                       nullopt,
                                       nullopt,
                                       nullopt,
                                       eps_abs,
                                       0,
                                       nullopt,
                                       nullopt,
                                       nullopt,
                                       nullopt,
                                       true,
                                       true,
                                       nullopt,
                                       initial_guess);
  T pri_res = std::max((qp.A * results.x - qp.b).lpNorm<Eigen::Infinity>(),
                       (helpers::positive_part(qp.C * results.x - qp.u) +
                        helpers::negative_part(qp.C * results.x - qp.l))
                         .lpNorm<Eigen::Infinity>());
  T dua_res = (qp.H * results.x + qp.g + qp.A.transpose() * results.y +
               qp.C.transpose() * results.z)
                .lpNorm<Eigen::Infinity>();
  DOCTEST_CHECK(pri_res <= eps_abs);
  DOCTEST_CHECK(dua_res <= eps_abs);

  std::cout << "------using API solving qp with dim: " << dim
            << " neq: " << n_eq << " nin: " << n_in << std::endl;
  std::cout << "primal residual: " << pri_res << std::endl;
  std::cout << "dual residual: " << dua_res << std::endl;
  std::cout << "total number of iteration: " << results.info.iter << std::endl;
  std::cout << "setup timing " << results.info.setup_time << " solve time "
            << results.info.solve_time << std::endl;
}

File: test/src/dense_qp_solve.py
#
# Copyright (c) 2022, INRIA
#
import os
import proxsuite
import numpy as np
import scipy.sparse as spa
import scipy.io as spio
import unittest


def normInf(x):
    if x.shape[0] == 0:
        return 0.0
    else:
        return np.linalg.norm(x, np.inf)


def generate_mixed_qp(n, seed=1):
    """
    Generate sparse problem in dense QP format
    """
    np.random.seed(seed)

    m = int(n / 4) + int(n / 4)
    # m  = n
    n_eq = int(n / 4)
    n_in = int(n / 4)

    P = spa.random(
        n, n, density=0.075, data_rvs=np.random.randn, format="csc"
    ).toarray()
    P = (P + P.T) / 2.0

    s = max(np.absolute(np.linalg.eigvals(P)))
    P += (abs(s) + 1e-02) * spa.eye(n)
    P = spa.coo_matrix(P)
    # print("sparsity of P : {}".format((P.nnz) / (n**2)))
    q = np.random.randn(n)
    A = spa.random(m, n, density=0.15, data_rvs=np.random.randn, format="csc").toarray(
        order="C"
    )  # row-major
    v = np.random.randn(n)  # Fictitious solution
    u = A @ v
    l = -1.0e20 * np.ones(m)

    return P.toarray(), q, A[:n_eq, :], u[:n_eq], A[n_in:, :], u[n_in:], l[n_in:]


class DenseQpWrapper(unittest.TestCase):
    # TESTS DENSE SOLVE FUNCTION

    def test_case_basic_solve(self):
        print(
            "------------------------sparse random strongly convex qp with equality and inequality constraints: test basic solve"
        )
        n = 10
        H, g, A, b, C, u, l = generate_mixed_qp(n)
        n_eq = A.shape[0]
        n_in = C.shape[0]

        results = proxsuite.proxqp.dense.solve(
            H=H,
            g=np.asfortranarray(g),
            A=A,
            b=np.asfortranarray(b),
            C=C,
            l=np.asfortranarray(l),
            u=np.asfortranarray(u),
            eps_abs=1.0e-9,
            eps_rel=0,
        )
        dua_res = normInf(
            H @ results.x + g + A.transpose() @ results.y + C.transpose() @ results.z
        )
        pri_res = max(
            normInf(A @ results.x - b),
            normInf(
                np.maximum(C @ results.x - u, 0) + np.minimum(C @ results.x - l, 0)
            ),
        )
        assert dua_res <= 1e-9
        assert pri_res <= 1e-9
        print("--n = {} ; n_eq = {} ; n_in = {}".format(n, n_eq, n_in))
        print("dual residual = {} ; primal residual = {}".format(dua_res, pri_res))
        print("total number of iteration: {}".format(results.info.iter))
        print(
            "setup timing = {} ; solve time = {}".format(
                results.info.setup_time, results.info.solve_time
            )
        )

    def test_case_different_rho_value(self):
        print(
            "------------------------sparse random strongly convex qp with equality and inequality constraints: test different rho values"
        )
        n = 10
        H, g, A, b, C, u, l = generate_mixed_qp(n)
        n_eq = A.shape[0]
        n_in = C.shape[0]

        results = proxsuite.proxqp.dense.solve(
            H=H,
            g=np.asfortranarray(g),
            A=A,
            b=np.asfortranarray(b),
            C=C,
            l=np.asfortranarray(l),
            u=np.asfortranarray(u),
            eps_abs=1.0e-9,
            eps_rel=0,
            rho=1.0e-7,
        )
        dua_res = normInf(
            H @ results.x + g + A.transpose() @ results.y + C.transpose() @ results.z
        )
        pri_res = max(
            normInf(A @ results.x - b),
            normInf(
                np.maximum(C @ results.x - u, 0) + np.minimum(C @ results.x - l, 0)
            ),
        )
        assert results.info.rho == 1e-7
        assert dua_res <= 1e-9
        assert pri_res <= 1e-9
        print("--n = {} ; n_eq = {} ; n_in = {}".format(n, n_eq, n_in))
        print("dual residual = {} ; primal residual = {}".format(dua_res, pri_res))
        print("total number of iteration: {}".format(results.info.iter))
        print(
            "setup timing = {} ; solve time = {}".format(
                results.info.setup_time, results.info.solve_time
            )
        )

    def test_case_different_mu_values(self):
        print(
            "------------------------sparse random strongly convex qp with equality and inequality constraints: test different mu_eq and mu_in values"
        )
        n = 10
        H, g, A, b, C, u, l = generate_mixed_qp(n)
        n_eq = A.shape[0]
        n_in = C.shape[0]

        results = proxsuite.proxqp.dense.solve(
            H=H,
            g=np.asfortranarray(g),
            A=A,
            b=np.asfortranarray(b),
            C=C,
            l=np.asfortranarray(l),
            u=np.asfortranarray(u),
            eps_abs=1.0e-9,
            eps_rel=0,
            mu_eq=1.0e-2,
            mu_in=1.0e-2,
        )
        dua_res = normInf(
            H @ results.x + g + A.transpose() @ results.y + C.transpose() @ results.z
        )
        pri_res = max(
            normInf(A @ results.x - b),
            normInf(
                np.maximum(C @ results.x - u, 0) + np.minimum(C @ results.x - l, 0)
            ),
        )
        assert dua_res <= 1e-9
        assert pri_res <= 1e-9
        print("--n = {} ; n_eq = {} ; n_in = {}".format(n, n_eq, n_in))
        print("dual residual = {} ; primal residual = {}".format(dua_res, pri_res))
        print("total number of iteration: {}".format(results.info.iter))
        print(
            "setup timing = {} ; solve time = {}".format(
                results.info.setup_time, results.info.solve_time
            )
        )

    def test_case_different_warm_starting(self):
        print(
            "------------------------sparse random strongly convex qp with equality and inequality constraints: test warm starting"
        )
        n = 10
        H, g, A, b, C, u, l = generate_mixed_qp(n)
        n_eq = A.shape[0]
        n_in = C.shape[0]
        x_wm = np.random.randn(n)
        y_wm = np.random.randn(n_eq)
        z_wm = np.random.randn(n_in)
        results = proxsuite.proxqp.dense.solve(
            H=H,
            g=np.asfortranarray(g),
            A=A,
            b=np.asfortranarray(b),
            C=C,
            l=np.asfortranarray(l),
            u=np.asfortranarray(u),
            eps_abs=1.0e-9,
            eps_rel=0,
            x=x_wm,
            y=y_wm,
            z=z_wm,
        )
        dua_res = normInf(
            H @ results.x + g + A.transpose() @ results.y + C.transpose() @ results.z
        )
        pri_res = max(
            normInf(A @ results.x - b),
            normInf(
                np.maximum(C @ results.x - u, 0) + np.minimum(C @ results.x - l, 0)
            ),
        )
        assert dua_res <= 1e-9
        assert pri_res <= 1e-9
        print("--n = {} ; n_eq = {} ; n_in = {}".format(n, n_eq, n_in))
        print("dual residual = {} ; primal residual = {}".format(dua_res, pri_res))
        print("total number of iteration: {}".format(results.info.iter))
        print(
            "setup timing = {} ; solve time = {}".format(
                results.info.setup_time, results.info.solve_time
            )
        )

    def test_case_different_verbose_true(self):
        print(
            "------------------------sparse random strongly convex qp with equality and inequality constraints: test verbose = true"
        )
        n = 10
        H, g, A, b, C, u, l = generate_mixed_qp(n)
        n_eq = A.shape[0]
        n_in = C.shape[0]
        results = proxsuite.proxqp.dense.solve(
            H=H,
            g=np.asfortranarray(g),
            A=A,
            b=np.asfortranarray(b),
            C=C,
            l=np.asfortranarray(l),
            u=np.asfortranarray(u),
            eps_abs=1.0e-9,
            eps_rel=0,
            verbose=True,
        )
        dua_res = normInf(
            H @ results.x + g + A.transpose() @ results.y + C.transpose() @ results.z
        )
        pri_res = max(
            normInf(A @ results.x - b),
            normInf(
                np.maximum(C @ results.x - u, 0) + np.minimum(C @ results.x - l, 0)
            ),
        )
        assert dua_res <= 1e-9
        assert pri_res <= 1e-9
        print("--n = {} ; n_eq = {} ; n_in = {}".format(n, n_eq, n_in))
        print("dual residual = {} ; primal residual = {}".format(dua_res, pri_res))
        print("total number of iteration: {}".format(results.info.iter))
        print(
            "setup timing = {} ; solve time = {}".format(
                results.info.setup_time, results.info.solve_time
            )
        )

    def test_case_different_no_initial_guess(self):
        print(
            "------------------------sparse random strongly convex qp with equality and inequality constraints: test no initial guess"
        )
        n = 10
        H, g, A, b, C, u, l = generate_mixed_qp(n)
        n_eq = A.shape[0]
        n_in = C.shape[0]
        results = proxsuite.proxqp.dense.solve(
            H=H,
            g=np.asfortranarray(g),
            A=A,
            b=np.asfortranarray(b),
            C=C,
            l=np.asfortranarray(l),
            u=np.asfortranarray(u),
            eps_abs=1.0e-9,
            eps_rel=0,
            initial_guess=proxsuite.proxqp.NO_INITIAL_GUESS,
        )
        dua_res = normInf(
            H @ results.x + g + A.transpose() @ results.y + C.transpose() @ results.z
        )
        pri_res = max(
            normInf(A @ results.x - b),
            normInf(
                np.maximum(C @ results.x - u, 0) + np.minimum(C @ results.x - l, 0)
            ),
        )
        assert dua_res <= 1e-9
        assert pri_res <= 1e-9
        print("--n = {} ; n_eq = {} ; n_in = {}".format(n, n_eq, n_in))
        print("dual residual = {} ; primal residual = {}".format(dua_res, pri_res))
        print("total number of iteration: {}".format(results.info.iter))
        print(
            "setup timing = {} ; solve time = {}".format(
                results.info.setup_time, results.info.solve_time
            )
        )

    def test_sparse_problem_with_exact_solution_known(self):
        print(
            "------------------------sparse random strongly convex qp with inequality constraints and exact solution known"
        )

        n = 150
        M = spa.lil_matrix(spa.eye(n))
        for i in range(1, n - 1):
            M[i, i + 1] = -1
            M[i, i - 1] = 1

        H = spa.csc_matrix(M.dot(M.transpose())).toarray()
        H = np.ascontiguousarray(H)
        g = -np.ones((n,))
        A = None
        b = None
        C = spa.csc_matrix(spa.eye(n)).toarray()
        C = np.ascontiguousarray(C)
        l = 2.0 * np.ones((n,))
        u = np.full(l.shape, +np.inf)

        results = proxsuite.proxqp.dense.solve(H, g, A, b, C, l, u)
        x_theoretically_optimal = np.array([2.0] * 149 + [3.0])

        dua_res = normInf(H @ results.x + g + C.transpose() @ results.z)
        pri_res = normInf(
            np.maximum(C @ results.x - u, 0) + np.minimum(C @ results.x - l, 0)
        )

        assert dua_res <= 1e-3  # default precision of the solver
        assert pri_res <= 1e-3
        assert normInf(x_theoretically_optimal - results.x) <= 1e-3
        print("--n = {} ; n_eq = {} ; n_in = {}".format(n, 0, n))
        print("dual residual = {} ; primal residual = {}".format(dua_res, pri_res))
        print("total number of iteration: {}".format(results.info.iter))
        print(
            "setup timing = {} ; solve time = {}".format(
                results.info.setup_time, results.info.solve_time
            )
        )

    def test_initializing_with_None(self):
        print("------------------------test initialization with Nones")

        H = np.array([[65.0, -22.0, -16.0], [-22.0, 14.0, 7.0], [-16.0, 7.0, 5.0]])
        g = np.array([-13.0, 15.0, 7.0])
        A = None
        b = None
        C = None
        _u = None
        _l = None

        results = proxsuite.proxqp.dense.solve(
            H,
            g,
            A,
            b,
            C,
        )
        print("optimal x: {}".format(results.x))

        dua_res = normInf(H @ results.x + g)

        assert dua_res <= 1e-3  # default precision of the solver
        print("--n = {} ; n_eq = {} ; n_in = {}".format(3, 0, 0))
        print("dual residual = {} ".format(dua_res))
        print("total number of iteration: {}".format(results.info.iter))
        print(
            "setup timing = {} ; solve time = {}".format(
                results.info.setup_time, results.info.solve_time
            )
        )

    def test_solve_qpsolvers_problem(self):
        print(
            "------------------------test case from qpsolvers with equality constraint and upper bound inequality constraints"
        )
        file_path = os.path.dirname(os.path.realpath(__file__))
        data_path = os.path.join(file_path, "..", "data")
        m = spio.loadmat(
            os.path.join(data_path, "simple_qp_with_inifinity_lower_bound.mat"),
            squeeze_me=True,
        )
        P = np.ascontiguousarray(m["P"].astype(float))
        q = m["q"].astype(float)
        A = np.ascontiguousarray(m["A"].astype(float).reshape((1, 3)))
        b = np.array([m["b"]]).reshape((1,))
        C = np.ascontiguousarray(m["C"].astype(float))
        l = m["l"].astype(float)
        u = m["u"].astype(float)

        results = proxsuite.proxqp.dense.solve(P, q, A, b, C, l, u, verbose=False)
        print("optimal x: {}".format(results.x))

        dua_res = normInf(
            P @ results.x + q + A.transpose() @ results.y + C.transpose() @ results.z
        )
        pri_res = max(
            normInf(A @ results.x - b),
            normInf(
                np.maximum(C @ results.x - u, 0) + np.minimum(C @ results.x - l, 0)
            ),
        )
        assert dua_res <= 1e-5
        assert pri_res <= 1e-5

        print("--n = {} ; n_eq = {} ; n_in = {}".format(3, 1, 3))
        print("dual residual = {} ".format(dua_res))
        print("total number of iteration: {}".format(results.info.iter))
        print(
            "setup timing = {} ; solve time = {}".format(
                results.info.setup_time, results.info.solve_time
            )
        )


if __name__ == "__main__":
    unittest.main()
